packages/stream-text/src/index.ts

```diff
@@ -32,96 +50,279 @@ export interface StreamTextOptions extends ChatOptions {
+            function: {
+              arguments: '',
+              name: fn.name,
+            },
+          } as StreamTextToolCall
+          toolCall.id = id
+          toolCall.type = type
+          toolCall.function.arguments += fn.arguments
+        }
+
+        /**
+         * check if toolcall is ended
+         */
+        const state = choiceState[index] ??= {
+          calledToolCallIDs: new Set(),
+          currentToolID: null,
+          endedToolCallIDs: new Set(),
+          index,
+          toolCallErrors: {},
+          toolCallResults: {},
+        }
+        // eslint-disable-next-line ts/strict-boolean-expressions
+        if (finish_reason) {
+          // end choice will end the current toolcall too
+          if (state.currentToolID !== null) {
+            endToolCall(state, state.currentToolID)
+          }
+        }
+        for (const toolCall of delta.tool_calls || []) {
+          // new toolcall
+          if (state.currentToolID !== null && state.currentToolID !== toolCall.id) {
+            endToolCall(state, state.currentToolID)
+          }
+
+          state.calledToolCallIDs.add(toolCall.id)
+          state.currentToolID = toolCall.id
+        }
+      },
+    })),
+    )
+
+    shouldOutputText && step.messages.push({
+      content: step.choices[0].message.content ?? '',
+      refusal: step.choices[0].message.refusal,
+      role: 'assistant',
+    })
+
+    // make actual toolcall and wait
+    await Promise.allSettled(step.choices.map(async (choice) => {
+      const state = choiceState[choice.index]
+
+      return Promise.allSettled([...state.endedToolCallIDs].map(async (id) => {
+        // eslint-disable-next-line ts/strict-boolean-expressions
+        if (state.toolCallErrors[id]) {
+          return
+        }
+
+        const toolCall = choice.message.tool_calls![id]
+        // eslint-disable-next-line sonarjs/no-nested-functions
+        const tool = options.tools?.find(tool => tool.function.name === toolCall.function.name)
+        if (tool) {
+          try {
+            const ret = await tool.execute(toolCall.function.parsed_arguments, {
+              abortSignal: options.abortSignal,
+              messages: options.messages,
+              toolCallId: id,
+            })
+            state.toolCallResults[id] = ret
+            step.messages.push({
+              content: ret,
+              role: 'tool',
+              tool_call_id: id,
+            })
+          }
+          catch (error) {
+            state.toolCallErrors[id] = error as Error
+          }
+        }
+        else {
+          state.toolCallErrors[id] = new Error(`tool ${toolCall.function.name} not found`)
+        }
+      }))
+    }))
+
+    steps.push(step)
+    stepCtrl?.enqueue(step)
+
+    options.onStepFinish?.(step)
+
+    if (shouldOutputText) {
+      return
     }
+
+    return async () => stepOne({ ...options, messages: step.messages })
   }
 
-  let buffer = ''
+  const main = async () => {
```

Comment:
```suggestion
  const invokeFunctionCalls = async () => {
```

The name of `main` is way too hard for others to understand immediately, I would suggest to have another name that includes the responsibility of the function here, perhaps `invokeFunctionCalls`? or `invokeForSteps`?